### PR TITLE
gh-111789: Use PyDict_GetItemRef() in _ctypes

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-111789

### CODE REVIEW

**Comment by vstinner** [comment]
Ping me when this PR is ready for review. Now it's marked as a draft.

**Comment by serhiy-storchaka** [comment]
It is ready for review. I marked it as draft because I do not insist that all these changes should be applied.

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -243,26 +243,13 @@ PyDict_SetItemProxy(PyObject *dict, PyObject *key, PyObject *item)
...
 _PyDict_GetItemProxy(PyObject *dict, PyObject *key, PyObject **presult)
 {
-    PyObject *item = PyDict_GetItemWithError(dict, key);
-    if (item == NULL) {
-        if (PyErr_Occurred()) {
-            return -1;
-        }
-        *presult = NULL;
-        return 0;
+    int rc = PyDict_GetItemRef(dict, key, presult);
```
Comment (vstinner) [inline Modules/_ctypes/_ctypes.c:246]: _PyDict_GetItemProxy() now returns 1 if PyDict_GetItemRef() returns 1 and PyWeakref_CheckProxy() is false. It's only called once with `if (_PyDict_GetItemProxy(cache, key, &result) != 0) {`. So it's ok.

File: Modules/_ctypes/_ctypes.c [hunk 2]
```diff
@@ -4870,15 +4861,12 @@ PyCArrayType_from_ctype(PyObject *itemtype, Py_ssize_t length)
         return NULL;
 
     PyObject *result;
-    if (_PyDict_GetItemProxy(cache, key, &result) < 0) {
-        Py_DECREF(key);
-        return NULL;
-    }
-    if (result) {
+    if (_PyDict_GetItemProxy(cache, key, &result) != 0) {
+        // error or found
```
Comment (vstinner) [inline Modules/_ctypes/_ctypes.c]:
```suggestion
        // found or error
```

File: Modules/_ctypes/_ctypes.c [hunk 3]
```diff
@@ -4870,15 +4861,12 @@ PyCArrayType_from_ctype(PyObject *itemtype, Py_ssize_t length)
...
         Py_DECREF(key);
         return result;
     }
-
+    // not found
```
Comment (vstinner) [inline Modules/_ctypes/_ctypes.c:4869]: thanks for adding this comment!

**Review by vstinner — APPROVED** [review]
LGTM

File: Modules/_ctypes/_ctypes.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_ctypes/_ctypes.c:246]: Yes, it has the same semantic as `PyDict_GetItemRef()`.

1 -- found, returns a strong reference
0 -- not found (or was destroyed)
-1 -- error

File: Modules/_ctypes/callproc.c [hunk 4]
```diff
@@ -1922,13 +1921,11 @@ create_pointer_type(PyObject *module, PyObject *cls)
...
     PyObject *key;
 
-    result = PyDict_GetItemWithError(_ctypes_ptrtype_cache, cls);
-    if (result) {
-        return Py_NewRef(result);
-    }
-    else if (PyErr_Occurred()) {
-        return NULL;
+    if (PyDict_GetItemRef(_ctypes_ptrtype_cache, cls, &result) != 0) {
+        // error or found
```
Comment (serhiy-storchaka) [inline Modules/_ctypes/callproc.c]:
```suggestion
        // found or error
```

**Pull request merged**